lib/phoenix_live_view/channel.ex

```diff
@@ -157,41 +213,138 @@ defmodule Phoenix.LiveView.Channel do
     """
   end
 
-  defp view_module(%{view: view}), do: view
+  defp view_module(%{socket: %Socket{view: view}}), do: view
+
+  defp decode("form", _router, url_encoded), do: Plug.Conn.Query.decode(url_encoded)
+  defp decode(_, _router, value), do: value
 
-  defp decode("form", url_encoded), do: Plug.Conn.Query.decode(url_encoded)
-  defp decode(_, value), do: value
+  defp handle_changed(state, %Socket{} = new_socket, ref, pending_internal_live_redirect \\ nil) do
+    new_state = %{state | socket: new_socket}
 
-  defp reply_render(state, socket, ref) do
-    case maybe_diff(state, socket) do
-      {diff, new_state} -> reply(new_state, ref, :ok, diff)
-      :noop -> reply(state, ref, :ok, %{})
+    case maybe_changed(new_socket) do
+      :diff ->
+        {diff, new_state} = render_diff(new_state, new_socket)
+
+        {:ok, :diff,
+         new_state
+         |> push_internal_live_redirect(pending_internal_live_redirect, nil)
+         |> push_render(diff, ref)}
+
+      :noop ->
+        {:ok, :noop,
+         state
+         |> push_internal_live_redirect(pending_internal_live_redirect, nil)
+         |> push_noop(ref)}
+
+      {:redirect, %{to: to} = opts} ->
+        send(new_state.transport_pid, {:socket_close, self(), {:redirect, to}})
+        {:stop, {:shutdown, {:redirect, to}}, push_redirect(new_state, opts, ref)}
+
+      {:live_redirect, {:internal, params}, %{to: _to, kind: _kind} = opts} ->
+        state
+        |> drop_redirect()
+        |> sync_handle_params_with_live_redirect(params, opts, ref)
+
+      {:live_redirect, :external, %{to: to} = opts} ->
+        send(state.transport_pid, {:socket_close, self(), {:redirect, to}})
+        opts = Map.put(opts, :flash, View.sign_flash(state.socket, View.get_flash(state.socket)))
+        {:stop, {:shutdown, {:redirect, to}}, push_external_live_redirect(state, opts, ref)}
     end
   end
 
-  defp push_render(state, socket) do
-    case maybe_diff(state, socket) do
-      {diff, new_state} -> push(new_state, "render", diff)
-      :noop -> state
+  defp drop_redirect(state) do
+    %{state | socket: View.drop_redirect(state.socket)}
+  end
+
+  defp sync_handle_params_with_live_redirect(state, params, opts, ref) do
+    if function_exported?(state.socket.view, :handle_params, 2) do
+      case state.socket.view.handle_params(params, state.socket) do
```

Comment:
We should change `handle_changed` to return `:noreply` and then you can just pipe this result into `handle_result`.